flytekit/extend/backend/base_agent.py

```diff
@@ -88,186 +178,176 @@ class AgentRegistry(object):
+        task_category = TaskCategory(name=task_type_name, version=task_type_version)
+        if task_category not in AgentRegistry._REGISTRY:
+            raise FlyteAgentNotFound(f"Cannot find agent for task category: {task_category}.")
+        return AgentRegistry._REGISTRY[task_category]
+
+    @staticmethod
+    def list_agents() -> List[Agent]:
+        return list(AgentRegistry.METADATA.values())
 
     @staticmethod
     def get_agent_metadata(name: str) -> Agent:
-        if name not in AgentRegistry._METADATA:
+        if name not in AgentRegistry.METADATA:
             raise FlyteAgentNotFound(f"Cannot find agent for name: {name}.")
-        return AgentRegistry._METADATA[name]
-
+        return AgentRegistry.METADATA[name]
 
-def mirror_async_methods(func: typing.Callable, **kwargs) -> typing.Coroutine:
-    if inspect.iscoroutinefunction(func):
-        return func(**kwargs)
-    args = [v for _, v in kwargs.items()]
-    return asyncio.get_running_loop().run_in_executor(None, func, *args)
 
-
-def convert_to_flyte_phase(state: str) -> TaskExecution.Phase:
-    """
-    Convert the state from the agent to the phase in flyte.
+class SyncAgentExecutorMixin:
     """
-    state = state.lower()
-    # timedout is the state of Databricks job. https://docs.databricks.com/en/workflows/jobs/jobs-2.0-api.html#runresultstate
-    if state in ["failed", "timeout", "timedout", "canceled"]:
-        return TaskExecution.FAILED
-    elif state in ["done", "succeeded", "success"]:
-        return TaskExecution.SUCCEEDED
-    elif state in ["running"]:
-        return TaskExecution.RUNNING
-    raise ValueError(f"Unrecognized state: {state}")
-
-
-def is_terminal_phase(phase: TaskExecution.Phase) -> bool:
-    """
-    Return true if the phase is terminal.
+    This mixin class is used to run the sync task locally, and it's only used for local execution.
+    Task should inherit from this class if the task can be run in the agent.
+
+    Synchronous tasks run quickly and can return their results instantly.
+    Sending a prompt to ChatGPT and getting a response, or retrieving some metadata from a backend system.
     """
-    return phase in [TaskExecution.SUCCEEDED, TaskExecution.ABORTED, TaskExecution.FAILED]
 
+    T = typing.TypeVar("T", "SyncAgentExecutorMixin", PythonTask)
+
+    def execute(self: T, **kwargs) -> LiteralMap:
+        from flytekit.tools.translator import get_serializable
+
+        ctx = FlyteContext.current_context()
+        ss = ctx.serialization_settings or SerializationSettings(ImageConfig())
+        task_template = get_serializable(OrderedDict(), ss, self).template
 
-def get_agent_secret(secret_key: str) -> str:
-    return flytekit.current_context().secrets.get(secret_key)
+        agent = AgentRegistry.get_agent(task_template.type, task_template.task_type_version)
+
+        resource = asyncio.run(self._do(agent, task_template, kwargs))
```

Comment:
Can we improve the error message here?
I have tested it in this PR.
https://github.com/flyteorg/flytekit/pull/2086
Here is the screenshot proof.
<img width="1168" alt="image" src="https://github.com/flyteorg/flytekit/assets/76461262/6ed81282-d80f-41a9-8588-8fac1bd321cd">

<img width="1165" alt="image" src="https://github.com/flyteorg/flytekit/assets/76461262/724e2de2-ac3b-405a-804a-3e4ebc378134">


```suggestion
        resource = asyncio.run(self._do(agent, task_template, kwargs))
        if resource.phase != TaskExecution.SUCCEEDED:
            raise FlyteUserException(f"Failed to run the task {agent.name} with error: {resource.message}")
```